### PR TITLE
V22F-180 replay driver

### PR DESCRIPTION
**Pull request by suppergerrie2**
Implement a driver that replays a recording

### CODE REVIEW

File: speed-dreams/source-2.2.3/src/drivers/replaydriver/Driver.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
+#include "Driver.h"
+#include "Mediator.h"
+#include "ConfigEnums.h"
+#include <tgf.h>
+#include <Recorder.h>
+
+#include "../../simulatedDrivingAssistance/rppUtils/RppUtils.hpp"
+
+#define RECORDING_NAME "userRecording20220426-140251"
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/src/drivers/replaydriver/Driver.cpp:9]: This is hardcoded for now, will be fixed in another PR

**Review by KungFuDonkey — APPROVED** [review]
The only thing I wonder is the define below, this might be a speed dreams must have, so I won't put it as a change request

File: speed-dreams/source-2.2.3/src/drivers/replaydriver/replaydriver.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
+#include <humandriver.h>
+#include <iostream>
+#include "Driver.h"
+
+static const int s_MaxBotAmount = 1;
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/drivers/replaydriver/replaydriver.cpp:5]: define?

File: speed-dreams/source-2.2.3/src/drivers/replaydriver/Driver.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        GfParmReadFile(std::experimental::filesystem::path(sdaFolder).append(CAR_SETTINGS_FILE_NAME).string().c_str(),
+                       GFPARM_RMODE_STD,
+                       true);
+
+    m_replayFile.open(sdaFolder.append(USER_INPUT_RECORDING_FILE_NAME).string().c_str());
+    SMediator::GetInstance()->RaceStart(p_track, p_carHandle, p_carParmHandle, p_situation);
+}
+
+/// @brief Start a new race.
+/// it opens te replay file and sets the time of the first input
```
Comment (noutheesink) [inline speed-dreams/source-2.2.3/src/drivers/replaydriver/Driver.cpp:42]:
```suggestion
/// it opens the replay file and sets the time of the first input
```

File: speed-dreams/source-2.2.3/src/drivers/replaydriver/Driver.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    m_replayFile >> inputTime;
+
+    if (m_replayFile.eof())
+    {
+        p_situation->raceInfo.state = RM_RACE_ENDED;
+    }
+    else
+    {
+        m_inputTime = std::stod(inputTime);
+    }
```
Comment (noutheesink) [inline speed-dreams/source-2.2.3/src/drivers/replaydriver/Driver.cpp:135]: Shouldn't you check for EOF before reading the new input time?
```suggestion
    if (m_replayFile.eof())
    {
        p_situation->raceInfo.state = RM_RACE_ENDED;
    }
    else
    {
        std::string inputTime;
        m_replayFile >> inputTime;
        m_inputTime = std::stod(inputTime);
    }
```

**Review by noutheesink — CHANGES_REQUESTED** [review]

**Review by noutheesink — APPROVED** [review]

**Pull request merged**